sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPartitionPlanner.scala

```diff
@@ -3,104 +3,292 @@
     requireNotNull(cosmosPartitioningConfig, "cosmosPartitioningConfig")
     require(defaultMaxPartitionSizeInMB >= 64, "Argument 'defaultMaxPartitionSizeInMB' must at least be 64")
     require(defaultMinimalPartitionCount >= 1, "Argument 'defaultMinimalPartitionCount' must at least be 1")
     //scalastyle:on multiple.string.literals
 
+    val planningInfo = this.getPartitionPlanningInfo(partitionMetadata, readLimit)
+
     cosmosPartitioningConfig.partitioningStrategy match {
       case PartitioningStrategies.Restrictive =>
-        applyRestrictiveStrategy(
-          cosmosClientConfig,
-          cosmosClientStateHandle,
-          cosmosContainerConfig
-        )
+        applyRestrictiveStrategy(planningInfo)
       case PartitioningStrategies.Custom =>
-        applyCustomStrategy(cosmosClientConfig,
-                            cosmosClientStateHandle,
-                            cosmosContainerConfig,
-                            changeFeedOffset,
-                            cosmosPartitioningConfig.targetedPartitionCount.get)
+        applyCustomStrategy(
+          container,
+          planningInfo,
+          cosmosPartitioningConfig.targetedPartitionCount.get)
       case PartitioningStrategies.Default =>
         applyStorageAlignedStrategy(
-          cosmosClientConfig,
-          cosmosClientStateHandle,
-          cosmosContainerConfig,
-          changeFeedOffset,
+          container,
+          planningInfo,
           1 / defaultMaxPartitionSizeInMB.toDouble,
           defaultMinimalPartitionCount
         )
       case PartitioningStrategies.Aggressive =>
         applyStorageAlignedStrategy(
-          cosmosClientConfig,
-          cosmosClientStateHandle,
-          cosmosContainerConfig,
-          changeFeedOffset,
-          3 / defaultMaxPartitionSizeInMB.toDouble,
+          container,
+          planningInfo,
+          5 / defaultMaxPartitionSizeInMB.toDouble,
           defaultMinimalPartitionCount
         )
     }
   }
 
-  private[this] def applyRestrictiveStrategy(
-      cosmosClientConfig: CosmosClientConfiguration,
-      cosmosClientStateHandle: Option[Broadcast[CosmosClientMetadataCachesSnapshot]],
-      cosmosContainerConfig: CosmosContainerConfig
-  ): Array[InputPartition] = {
-    val feedRangesList = this
-      .getFeedRanges(cosmosClientConfig,
-                     cosmosClientStateHandle,
-                     cosmosContainerConfig)
+  private[this] def getContinuationTokenLsnOfFirstItem(items: Iterable[ObjectNode]): Option[String] = {
+    items
+      .collectFirst({
+        case item: ObjectNode if item != null =>
+          val lsnNode = item.get(LsnAttributeName)
+          if (lsnNode != null && lsnNode.isNumber) {
+            // when grabbing the LSN from the item we need to use teh item's LSN -1
+            // to ensure we would retrieve this item again
+            Some(
+              SparkBridgeImplementationInternal.toContinuationToken(lsnNode.asLong() - 1))
```

Comment:
You are probably right, I have not tested it, but is the `-1` really required? I know it's an approximation as that `-1` might be including 1 document or 100 because it's the transaction number.

Reply:
Yes - tests cover this - It also isn't an approximation. Basically this is used to map StartFromSettings to LSN.

When we set StartFromPointInTime or From Beginning we get back some records - to identify the LSN that we would need to specify in an artificially created continuation token we need to use an LSN of the first returned document -1. Backend will return all records > the LSN of continuation token - so we get back the documents again

Reply:
Does that make sense?